src/controllers/bookController.ts

```diff
@@ -0,0 +1,128 @@
+import { NextFunction, Request, Response } from "express";
+import Book, { IBook } from "../models/Book";
+import { StatusCodes } from "http-status-codes";
+import { BadRequestError, NotFoundError } from "../errors";
+import mongoose from "mongoose";
+
+// Interface for query parameters
+interface IGetBooksQuery {
+  search?: string;
+  categories?: string;
+  sort?: "a-z" | "z-a" | "latest" | "oldest";
+}
+
+// Get all books
+const getAllBooks = async (
+  req: Request<{}, {}, {}, IGetBooksQuery>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { search, categories, sort } = req.query;
+  let query: any = {};
+
+  // Filter by categories
+  if (categories) {
+    const categoryArray = Array.isArray(categories)
+      ? categories
+      : categories.split(",");
+
+    query.categories = { $in: categoryArray };
+  }
+
+  // Search by title
+  if (search) {
+    query.title = { $regex: new RegExp(search, "i") }; // Case-insensitive search
+  }
+
+  try {
+    // Sort options
+    const sortOptions: any = {};
+    if (sort) {
+      if (sort === "a-z") sortOptions.title = 1;
+      else if (sort === "z-a") sortOptions.title = -1;
+      else if (sort === "latest") sortOptions.publishedDate = -1;
+      else if (sort === "oldest") sortOptions.publishedDate = 1;
+    } else {
+      sortOptions.publishedDate = -1; 
+    }
+
+    const books = await Book.find(query).sort(sortOptions);
+    res.status(StatusCodes.OK).json({ books, count: books.length });
+  } catch (error: unknown) {
+    return next(error);
+  }
+};
+
+// Get a single book by ID
+const getBook = async (
+  req: Request<{ bookId: string }>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { bookId } = req.params;
+
+  try {
+     
+    const book = await Book.findById(bookId);
+    if (!book) {
+      return next(new NotFoundError(`No book with id ${bookId}`));
```

Comment:
This is not a pattern I've seen very often to use a global handler for errors.

Rather, I would expect you to explicitly send a response with an error code and a message in-line here.

```
res.status(404).json({error: `no book with id ${bookId}`});
return;
```

Why should I have to go to another file and try to decipher based on the error type I pass here what status code should be returned? The status code information is really important for the client to decide what went wrong and whose fault it was.

I see that you've defined custom error types with the name corresponding to the status code being returned. All they do is set a special status code on the response, which just feels like unnecessary extra code to me.

Ultimately, I think this is my personal taste speaking and so I will leave this to you to decide to keep or remove.